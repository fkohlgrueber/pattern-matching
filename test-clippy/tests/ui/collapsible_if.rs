// Copyright 2014-2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#[rustfmt::skip]
#[warn(clippy::collapsible_if)]
fn main() {
    let x = "hello";
    let y = "world";
    if x == "hello" {
        if y == "world" {
            println!("Hello world!");
        }
    }

    if x == "hello" || x == "world" {
        if y == "world" || y == "hello" {
            println!("Hello world!");
        }
    }

    if x == "hello" && x == "world" {
        if y == "world" || y == "hello" {
            println!("Hello world!");
        }
    }

    if x == "hello" || x == "world" {
        if y == "world" && y == "hello" {
            println!("Hello world!");
        }
    }

    if x == "hello" && x == "world" {
        if y == "world" && y == "hello" {
            println!("Hello world!");
        }
    }

    if 42 == 1337 {
        if 'a' != 'A' {
            println!("world!")
        }
    }

    // Collapse `else { if .. }` to `else if ..`
    if x == "hello" {
        print!("Hello ");
    } else {
        if y == "world" {
            println!("world!")
        }
    }

    if x == "hello" {
        print!("Hello ");
    } else {
        if let Some(42) = Some(42) {
            println!("world!")
        }
    }

    if x == "hello" {
        print!("Hello ");
    } else {
        if y == "world" {
            println!("world")
        }
        else {
            println!("!")
        }
    }

    if x == "hello" {
        print!("Hello ");
    } else {
        if let Some(42) = Some(42) {
            println!("world")
        }
        else {
            println!("!")
        }
    }

    if let Some(42) = Some(42) {
        print!("Hello ");
    } else {
        if let Some(42) = Some(42) {
            println!("world")
        }
        else {
            println!("!")
        }
    }

    if let Some(42) = Some(42) {
        print!("Hello ");
    } else {
        if x == "hello" {
            println!("world")
        }
        else {
            println!("!")
        }
    }

    if let Some(42) = Some(42) {
        print!("Hello ");
    } else {
        if let Some(42) = Some(42) {
            println!("world")
        }
        else {
            println!("!")
        }
    }

    // Works because any if with an else statement cannot be collapsed.
    if x == "hello" {
        if y == "world" {
            println!("Hello world!");
        }
    } else {
        println!("Not Hello world");
    }

    if x == "hello" {
        if y == "world" {
            println!("Hello world!");
        } else {
            println!("Hello something else");
        }
    }

    if x == "hello" {
        print!("Hello ");
        if y == "world" {
            println!("world!")
        }
    }
}
