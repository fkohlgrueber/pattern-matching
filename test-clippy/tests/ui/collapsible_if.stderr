error: this if statement can be collapsed
  --> $DIR/collapsible_if.rs:15:5
   |
LL | /     if x == "hello" {
LL | |         if y == "world" {
LL | |             println!("Hello world!");
LL | |         }
LL | |     }
   | |_____^
   |
   = note: #[forbid(simple_pattern)] on by default

error: this if statement can be collapsed
  --> $DIR/collapsible_if.rs:21:5
   |
LL | /     if x == "hello" || x == "world" {
LL | |         if y == "world" || y == "hello" {
LL | |             println!("Hello world!");
LL | |         }
LL | |     }
   | |_____^

error: this if statement can be collapsed
  --> $DIR/collapsible_if.rs:27:5
   |
LL | /     if x == "hello" && x == "world" {
LL | |         if y == "world" || y == "hello" {
LL | |             println!("Hello world!");
LL | |         }
LL | |     }
   | |_____^

error: this if statement can be collapsed
  --> $DIR/collapsible_if.rs:33:5
   |
LL | /     if x == "hello" || x == "world" {
LL | |         if y == "world" && y == "hello" {
LL | |             println!("Hello world!");
LL | |         }
LL | |     }
   | |_____^

error: this if statement can be collapsed
  --> $DIR/collapsible_if.rs:39:5
   |
LL | /     if x == "hello" && x == "world" {
LL | |         if y == "world" && y == "hello" {
LL | |             println!("Hello world!");
LL | |         }
LL | |     }
   | |_____^

error: this if statement can be collapsed
  --> $DIR/collapsible_if.rs:45:5
   |
LL | /     if 42 == 1337 {
LL | |         if 'a' != 'A' {
LL | |             println!("world!")
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:54:12
   |
LL |       } else {
   |  ____________^
LL | |         if y == "world" {
LL | |             println!("world!")
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:62:12
   |
LL |       } else {
   |  ____________^
LL | |         if let Some(42) = Some(42) {
LL | |             println!("world!")
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:70:12
   |
LL |       } else {
   |  ____________^
LL | |         if y == "world" {
LL | |             println!("world")
LL | |         }
...  |
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:81:12
   |
LL |       } else {
   |  ____________^
LL | |         if let Some(42) = Some(42) {
LL | |             println!("world")
LL | |         }
...  |
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:92:12
   |
LL |       } else {
   |  ____________^
LL | |         if let Some(42) = Some(42) {
LL | |             println!("world")
LL | |         }
...  |
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:103:12
   |
LL |       } else {
   |  ____________^
LL | |         if x == "hello" {
LL | |             println!("world")
LL | |         }
...  |
LL | |         }
LL | |     }
   | |_____^

error: this `else { if .. }` block can be collapsed
  --> $DIR/collapsible_if.rs:114:12
   |
LL |       } else {
   |  ____________^
LL | |         if let Some(42) = Some(42) {
LL | |             println!("world")
LL | |         }
...  |
LL | |         }
LL | |     }
   | |_____^

error: aborting due to 13 previous errors

